### PR TITLE
[ROCm] add hipMAGMA support

### LINKED ISSUE #48831: Enable linear algebra functions on ROCm platform
## 🚀 Feature
PyTorch shipping with linear algebra functionality on AMD GPUs via the [hipMAGMA](https://icl.utk.edu/magma/software/browse.html?start=0&per=5) library. The ability to run the same code that uses linalg functions on Nvidia and AMD GPUs.

## Motivation

Dense linear algebra is used in many applications of scientific computing. In the High-Performance Computing world, AMD is gaining popularity with the announcements that the future world's fastest supercomputers will be based on AMD GPUs for example [Frontier and El Capitan](https://www.amd.com/en/products/exascale-era), [Finnish LUMI](https://www.lumi-supercomputer.eu/may-we-introduce-lumi/) and probably some others. This will bring many researchers using PyTorch + ROCm.

## Pitch

Currently, most of the linear algebra functionality in PyTorch relies on the MAGMA library, which is linked only in CUDA build.
For ROCm build, it's possible to link to hipMAGMA (a version of MAGMA that was processed with hipify script) and with minimal changes to the existing code all the linear algebra functions that call MAGMA will work for AMD GPUs as well.

`magma-hip` would need to be added to [pytorch/builder](https://github.com/pytorch/builder/tree/master/magma).
Some changes in code would need to be made so that only magma v2 API  `magma_v2.h` is used (maybe v1 works fine as well, but it includes only `cublas.h` by default, while v2 seems to work with `hipblas.h`.
And a few changes to CMake config files to enable searching and linking of MAGMA in ROCm build path.

## Alternatives

AMD develops its variant of Nvidia's cuSOLVER, called [rocSOLVER](https://github.com/ROCmSoftwarePlatform/rocSOLVER). So the alternative option would be to use rocSOLVER as the underlying library for linear algebra functionality.

## Additional context

I have compiled hipMAGMA myself for AMD Vega 20 (gfx906) and managed to link it with PyTorch so that SVD runs fine on AMD GPUs. I didn't test other functions.

cc @malfet @seemethere @walterddr @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @jeffdaily @sunway513 @stomov

### PR DESCRIPTION
**Pull request by jeffdaily**
Fixes #48831.  

- CI image is updated to build hipMAGMA from source and set env MAGMA_HOME.
- CMake is updated to separate different requirements for CUDA versus ROCm MAGMA.
- Some unit tests that become enabled with MAGMA are currently skipped for ROCm due to failures.  Fixing these failures will be follow-on work.

### CODE REVIEW

**Comment by jeffdaily** [comment]
ROCm CI puts us in a chicken and egg sort of situation.  This PR updates the CI image in addition to updating source files, tests, etc.  The ROCm CI Jenkins process uses a static image tag.  There's no clean way to actually test this PR; initial CI runs will use an older CI image that does not have MAGMA and will therefore appear to pass all tests.

**Comment by jeffdaily** [comment]
Rebased against upstream master to try and fix unrelated CI failures.

File: .circleci/docker/common/install_rocm.sh [hunk 1]
```diff
@@ -44,9 +37,23 @@ install_ubuntu() {
...
-  # Cleanup
-  apt-get autoclean && apt-get clean
-  rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+    # "install" hipMAGMA into /opt/rocm/magma by copying after build
+    git clone https://bitbucket.org/icl/magma.git -b hipMAGMA
+    pushd magma
+    cp make.inc-examples/make.inc.hip-mkl-gcc make.inc
+    echo 'LIBDIR += -L$(MKLROOT)/lib' >> make.inc
+    echo 'LIB += -Wl,--enable-new-dtags -Wl,--rpath,/opt/rocm/lib -Wl,--rpath,$(MKLROOT)/lib -Wl,--rpath,/opt/rocm/magma/lib' >> make.inc
+    echo 'DEVCCFLAGS += --amdgpu-target=gfx803 --amdgpu-target=gfx900 --amdgpu-target=gfx906 --amdgpu-target=gfx908' >> make.inc
```
Comment (malfet) [inline .circleci/docker/common/install_rocm.sh:46]: Hmm, can we infer supported gpu targets from environment variable or something?

**Review by malfet — APPROVED** [review]
LGTM, but can you please file issue and mention it in `@skipCUDAIfRocm`decorator, in order not to loose track about the reason why the test is skipped

File: .circleci/docker/common/install_rocm.sh (hunk 1 above)
Comment (jeffdaily) [reply .circleci/docker/common/install_rocm.sh:46]: We don't currently have any way of querying the default set of targets for a particular ROCm release.  Today, every ROCm package is specifying their own set of gfx targets, though we do coordinate across projects to ensure they match.  [Example from HIP](https://github.com/ROCm-Developer-Tools/HIP/blob/2080cc113a2d767352b512b9d24c0620b6dee790/CMakeLists.txt#L370).

This PR matches that current list.  As we upgrade ROCm releases and prepare PRs for new ROCm versions, this list can change.

We're also discussing internally whether we can provide a hipMAGMA package alongside our other ROCm packages.  Such a package doesn't exist yet, so we're forced to build from source today.

**Comment by jeffdaily** [comment]
The `@skipCUDAIfRocm` decorator doesn't allow a specific text description; adding issue as a comment instead.

**Comment by jeffdaily** [comment]
@malfet I added 1 comment above the first skipped test.  Let me know if you had something else in mind.  Also moved the magma install into a separate function so it could be used for both ubuntu and centos images.

**Comment by jeffdaily** [comment]
Rebased against `viable/strict`.

**Pull request closed without merging**